### DISCUSSION TITLE
Toys vs. Thor

### DISCUSSION THREAD

**Discussion by DArrigoni** [dazuma/toys#177] [General]
Just out of curiosity, how would you compare `toys` to `thor` (http://whatisthor.com/)? At first blush, they seem similar but I'm sure that there's probably some fundamental difference that separates the two.

**Comment by AlexWayfer** [comment]
I'm not an expert, but I've used a lot of similar tools and gems.

Let me try to explain:

1. There is a niche for local (development, simple) tasks, like tests, linting, generating, migration — and Rake is obviously primal/general here.
2. And there is a niche for CLI tools, like you create a gem with external Command Line Interface. And Thor is very popular here.

Then my opinion and experience (without negative):

1. Toys tries to move Rake from its throne, and this is great, because Rake had almost no alternatives, and there are already movements and rises like https://makefile.site/, but Toys has great features, simplicity and support (thank you, @dazuma).
2. [Clamp](https://github.com/mdub/clamp) seems more feature-complete and simple for me than Thor, from the point of modular CLI with nested commands, options, etc.

As examples:
* https://github.com/AlexWayfer/flame-cli has external CLI via `clamp` and development tools/commands/toys via `toys`.
* https://github.com/filewatcher/filewatcher-cli there you can see an evolution from the `Optimist` gem (was calling another, unappreciated).

Toys are also (pretty) easy extendable, so you can create repo/gems like https://github.com/AlexWayfer/gem_toys, without external CLI, but with possibility to re-use your Toys template/commands. It's great.

Then… your choice. You can do anything via anything. I hope I've made my point of view.

**Comment by dazuma** [comment]
I think @AlexWayfer channeled my thinking really well.

While Thor does have a `thor` executable and a `Thorfile` mechanism, it seems principally designed for the use case of making your own CLI executable. Nearly all the major community uses that I know (for example, Rails) are along those lines, and most of its features (and plugins and community) center around those use cases.

While Toys does provide classes letting you write your own executable, it was principally designed to target the Rake use cases, using the `toys` executable and writing your own project-specific `.toys` directories and `.toys.rb` files. Most of its interesting features focus on things that I wish Rake did, like command line options, global commands, bundler integration, online help, and test infrastructure.
